uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.uima.cas.impl;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static org.apache.uima.cas.SerialFormat.*;
+import static org.apache.uima.util.CasCreationUtils.createCas;
+import static org.apache.uima.util.CasIOUtils.load;
+import static org.apache.uima.util.CasIOUtils.save;
+import static org.junit.Assert.assertTrue;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.apache.uima.UIMAFramework;
+import org.apache.uima.cas.CAS;
+import org.apache.uima.cas.SerialFormat;
+import org.apache.uima.cas.text.AnnotationFS;
+import org.apache.uima.resource.metadata.TypeSystemDescription;
+import org.junit.Test;
+
+public class ConcurrentBinarySerializationTest
+{
+    /**
+     * Serialization of the CAS is not inherently thread-safe. This test tries to run multiple
+     * serializations of the CAS in parallel to trigger a situation where an invalid serialization
+     * is generated - to be found by deserializing again.
```

Comment:
I'm confused.  After the "synch" additions, serialization of the CAS should be thread-safe for multiple concurrent serializations.  But serialization of the CAS while another thread is updating the CAS (adding/modifying Feature Structures or adding/removing to/from indexes) is not safe (and is not currently "checked").

I assume that's what you meant to say?  This test should fail (before the sync fixes were added) and succeed after them?

Reply:
Right - it should fail right now as the PR is based on commit 1fb339310e7a11cae5398f0cb8a2ec2eb636905e which doesn't include your synchronization yet. But when merging in your fixes from master, it should work. What surprises me is that the build doesn't seem to have failed...

Reply:
Ah, Jenkins seems to temporarily merge the PR into master before building:

```
Commit message: "Merge e3cd9827a1424caeda481588f836d3a64e63816b into 382b096ae3967af6fa6fd6758fd36ae427372c6a"
```

Reply:
Anyway - if you are good with the test, we could merge it.

Reply:
Great, Thanks for this very nice test case :-) .   I wonder if it works in uima v2?